docs/posts/1tbc/index.qmd

```diff
@@ -0,0 +1,733 @@
+
+Let's explore the largest table, `lineitem`. This table in memory is ~6X larger
+than RAM.
+
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to explore the lineitem table"
+sf = 1
+n = 1
+data_dir = f"tpch_data/parquet/sf={sf}/n={n}"
+
+table_name = "lineitem"
+data = f"{data_dir}/{table_name}/*.parquet"
+
+t = ibis.read_parquet(data)
+print(f"rows: {t.count().to_pyarrow().as_py():,} | columns: {len(t.columns)}")
+```
+
+Over 6 billion rows!
+
+Let's try to display the first few rows with Ibis, pandas, and Polars:
+
+::: {.panel-tabset}
+
+## Ibis
+
+```{python}
+t = ibis.read_parquet(data)
+t.head(3)
+```
+
+## pandas
+
+```{.python}
+df = pd.concat([pd.read_parquet(f) for f in glob.glob(data)], ignore_index=True) # <1>
+df.head(3)
+```
+
+1. Work around lack of reading multiple parquet files in pandas
+
+```html
+The Kernel crashed while executing code in the current cell or a previous cell.
+Please review the code in the cell(s) to identify a possible cause of the failure.
+Click here for more info.
+View Jupyter log for further details.
+```
+
+## Polars (eager)
+
+```{.python}
+df = pl.read_parquet(data)
+df.head(3)
+```
+
+```html
+The Kernel crashed while executing code in the current cell or a previous cell.
+Please review the code in the cell(s) to identify a possible cause of the failure.
+Click here for more info.
+View Jupyter log for further details.
+```
+
+## Polars (lazy)
+
+```{python}
+df = pl.scan_parquet(data)
+df.head(3).collect()
+```
+
+## Polars (lazy, streaming)
+
+```{python}
+df = pl.scan_parquet(data)
+df.head(3).collect(streaming=True)
+```
+
+:::
+
+Ibis (DuckDB backend by default) can display the first few rows. Polars (lazy)
+can handle this in regular and streaming mode. pandas and Polars (eager) crash
+the kernel as they must load all the data into memory to construct their
+dataframes. Since this table is at ~6X larger than our 96GiB of RAM, we can't
+load it all.
+
+For lazily computed Python dataframes, an underlying query engine can determine
+the potential subset of data that needs to be read into memory for a given
+query. For example, to display any three rows from the `lineitem` table it can
+just read the first three rows from the first Parquet file in the dataset.
+
+:::{.callout-tip title="Visualize the Ibis expression tree" collapse="true"}
+
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to visualize the Ibis expression tree"
+from ibis.expr.visualize import to_graph
+
+to_graph(t.head(3))
+```
+
+:::
+
+Let's try something more challenging: sorting the `lineitem` table. This forces
+at least some columns from all rows of data to pass through the query engine to
+determine the top 3 rows per the specified ordering. Since the data is larger
+than RAM, only "streaming" engines can handle this. We'll try with the methods
+that worked on the previous query and add in the DataFusion backend for Ibis.
```

Comment:
This is an example of a place where [partial sorting](https://en.wikipedia.org/wiki/Partial_sorting) is useful (you only need to put the first K elements in order, not all N), not if that's worth linking to or not.